lib/rules/prefer-object-spread.js

```diff
@@ -0,0 +1,259 @@
+/**
+ * @fileoverview Prefers object spread property over Object.assign
+ * @author Sharmila Jesupaul
+ * See LICENSE file in root directory for full license.
+ */
+
+"use strict";
+
+const matchAll = require("string.prototype.matchall");
+
+/**
+ * Helper that checks if the node is an Object.assign call
+ * @param {ASTNode} node - The node that the rule warns on
+ * @returns {boolean} - Returns true if the node is an Object.assign call
+ */
+function isObjectAssign(node) {
+    return (
+        node.callee &&
+        node.callee.type === "MemberExpression" &&
+        node.callee.object.name === "Object" &&
+        node.callee.property.name === "assign"
+    );
+}
+
+/**
+ * Helper that checks if the node needs parentheses to be valid JS.
+ * The default is to wrap the node in parentheses to avoid parsing errors.
+ * @param {ASTNode} node - The node that the rule warns on
+ * @returns {boolean} - Returns true if the node needs parentheses
+ */
+function needsParens(node) {
+    const parent = node.parent;
+
+    if (!parent || !node.type) {
+        return true;
+    }
+
+    switch (parent.type) {
+        case "VariableDeclarator":
+        case "ArrayExpression":
+        case "ReturnStatement":
+        case "CallExpression":
+        case "Property":
+            return false;
+        default:
+            return true;
+    }
+}
+
+/**
+ * Determines if an argument needs parentheses. The default is to not add parens.
+ * @param {ASTNode} node - The node to be checked.
+ * @returns {boolean} True if the node needs parentheses
+ */
+function argNeedsParens(node) {
+    if (!node.type) {
+        return false;
+    }
+
+    switch (node.type) {
+        case "AssignmentExpression":
+        case "ArrowFunctionExpression":
+        case "ConditionalExpression":
+            return true;
+        default:
+            return false;
+    }
+}
+
+/**
+ * Helper that adds a comma after the last non-whitespace character that is not a part of a comment.
+ * @param {string} arg - String of argument text
+ * @returns {string} - argument with comma at the end of it
+ */
+function addComma(arg) {
+    const nonWhitespaceCharacterRegex = /[^\s\\]/g;
+    const commentRegex = /(\/\*[\w'\s\r\n*]*\*\/)|(\/\/[\w\s']*)|(<![-\-\s\w>/]*>)/g;
```

Comment:
Rather than taking a string and matching for comments, would it make more sense to use token/comment retrieval methods in `SourceCode`? I think this basically boils down to getting the last non-comment token in the argument node, and adding a comma afterward.

Reply:
Thank you for mentioning this. Instead of the regex, we can just use the ranges in `context.getCommentsInside(node)`. Definitely over complicated this a bit 😬